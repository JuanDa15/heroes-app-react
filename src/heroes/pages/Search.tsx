import { useEffect, useState } from 'react'
import Alert from '../../shared/components/Alert'
import SearchBar from '../../shared/components/Searchbar'
import HeroesList from '../components/HeroesList'
import { useLocation } from 'react-router-dom'
import getHeroesByName from '../helpers/getHeroesByName'
import { type Hero } from '../../constants'

export default function Search (): JSX.Element {
  const location = useLocation()
  const [heroes, setHeroes] = useState<Hero[]>([])

  const q = location.search.slice(location.search.lastIndexOf('=') + 1)
  const formattedQ = typeof q === 'string' ? q : ''

  useEffect(() => {
    void searchHeroes(formattedQ)
  }, [q])

  const searchHeroes = async (query: string): Promise<void> => {
    const resp = await getHeroesByName(query)
    setHeroes(resp)
  }

  const NO_RESULTS = ((heroes.length === 0) && (formattedQ.length > 0))
  const INFORMATIVE = ((heroes.length === 0) && (formattedQ.length === 0))
  return (
    <>
      {
        Boolean(formattedQ) && <h1 className="text-3xl mt-3">Search: <b>{formattedQ}</b></h1>
      }
      <hr className="h-px my-8 bg-gray-200 border-0 dark:bg-gray-700" />
      <SearchBar q={formattedQ}/>
      <hr className="h-px my-8 bg-gray-200 border-0 dark:bg-gray-700" />
      {
        NO_RESULTS && (
          <Alert type='Error' data-testid="error-message">
            No results found for {formattedQ}
          </Alert>
        )
      }
      {
        INFORMATIVE && (
          <Alert type='Info' data-testid="default-message">
            Search a hero
          </Alert>
        )
      }
      <HeroesList heroes={heroes} />
    </>
  )
}
