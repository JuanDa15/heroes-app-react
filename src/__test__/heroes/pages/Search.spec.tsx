import { MemoryRouter } from 'react-router-dom'
import Search from '../../../heroes/pages/Search'
import { render, screen } from '@testing-library/react'

describe('Test in <Search /> component', () => {
  test('Should render default value', () => {
    render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    )
    expect(screen.getAllByTestId('default-message')).toBeTruthy()
  })
  test('Should match with snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })
  test('Should show batman and the input with the same value', () => {
    const query = 'batman'
    render(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <Search />
      </MemoryRouter>
    )

    const input: HTMLInputElement = screen.getByTestId('search-input')
    screen.debug()
    expect(input.value).toBe(query)
  })
})
