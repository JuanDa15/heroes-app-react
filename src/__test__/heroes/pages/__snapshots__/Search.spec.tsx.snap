// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Search /> component Should match with snapshot 1`] = `
<div>
  <hr
    class="h-px my-8 bg-gray-200 border-0 dark:bg-gray-700"
  />
  <form>
    <label
      class="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
      for="default-search"
    >
      Search
    </label>
    <div
      class="relative"
    >
      <div
        class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-gray-500 dark:text-gray-400"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <input
        class="block w-full p-4 pl-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        data-testid="search-input"
        id="search"
        name="query"
        placeholder="Search Batman, Superman..."
        type="search"
        value=""
      />
      <button
        class="text-white absolute right-2.5 bottom-2.5 bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
        data-testid="search-button"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
  <hr
    class="h-px my-8 bg-gray-200 border-0 dark:bg-gray-700"
  />
  <div
    class="p-4 mb-4 text-sm rounded-lg text-blue-800 bg-blue-50 dark:bg-gray-800 dark:text-blue-400"
    data-testid="default-message"
    role="alert"
  >
    Search a hero
  </div>
  <ul
    class="grid grid-cols-2 md:grid-cols-3 gap-4 p-2"
  />
</div>
`;
